ouroboros-consensus-test/bench-storage/Bench/Ouroboros/Consensus/Storage/LedgerDB/HD.hs

```diff
@@ -0,0 +1,412 @@
+  still "warming up", i.e., when the length of the diff sequence is still less
+  than @k@: before that, flushes will not have any effect. To be precise, the
+  warmup phase corresponds to running the simulation/generator that generates
+  Push commands for exactly @k@ steps and throwing away the commands we generate
+  in the process.
+
+  We use the commands that we generate after the warmup phase as inputs to the
+  benchmarks, as well as the state of the diff sequence right after the warmup
+  phase. Convince yourself that interpretation of these commands only make sense
+  with respect to the "warmed up" diff sequence.
+-}
+module Bench.Ouroboros.Consensus.Storage.LedgerDB.HD (benchmarks) where
+
+import           Control.DeepSeq
+import           Control.Monad
+import           Control.Monad.RWS
+import           Data.ByteString.Short (ShortByteString)
+import qualified Data.ByteString.Short as B
+import           Data.Foldable
+import           Data.Kind (Type)
+import           GHC.Generics (Generic)
+
+import           QuickCheck.GenT
+import           Test.QuickCheck hiding (Result, choose, frequency, getSize,
+                     resize, shuffle, sized, variant)
+import           Test.Tasty.Bench
+import           Test.Tasty.QuickCheck hiding (choose, frequency, resize, sized,
+                     variant)
+
+import           Test.Util.MockDiffSeq
+
+
+
+benchmarks :: Benchmark
+benchmarks = bgroup "HD" [ bgroup "DiffSeq/Diff operations" [
+        env (mkBenchEnv gcDefault) $
+          \ ~bEnv@BenchEnv {cmdsMock, ds0Mock } ->
+            bgroup "Mocked diffs" [
+                bench "Benchmark interpret" $
+                  nf (interpretMock ds0Mock) cmdsMock
+              , testProperty "Sanity checks" $
+                  once (sanityCheck bEnv)
+              ]
+    ]]
+  where
+    mkBenchEnv :: GenConfig -> IO (BenchEnv k v)
+    mkBenchEnv conf = do
+      CmdGenResult{cmds, ds0, finalModel} <- generate $ generator conf
+      pure $ BenchEnv {
+          cmdsMock = cmds
+        , ds0Mock  = ds0
+        , model    = finalModel
+        }
+
+    interpretMock = interpret @MockDiffSeq @Key @Value
+
+-- | Inputs to a benchmark.
+data BenchEnv k v = BenchEnv {
+    cmdsMock :: ![Cmd MockDiffSeq k v]
+    -- | State of the diff sequence after the warmup phase during command
+    -- generation.
+  , ds0Mock  :: !(MockDiffSeq k v)
+    -- | Final state of the model after the warmup phase /and/
+    -- the command generation phase.
+  , model    :: !(Model k v)
+  }
+  deriving stock Generic
+  deriving anyclass NFData
+
+{-------------------------------------------------------------------------------
+  Commands
+-------------------------------------------------------------------------------}
+
+type Block (ds :: DiffSeqKind) k v = (SlotNo ds, Diff ds k v)
+
+data Cmd (ds :: DiffSeqKind) k v =
+    Push     (SlotNo ds)     (Diff ds k v)
+  | Flush    Int
+  | Rollback Int             [Block ds k v]
```

Comment:
I wonder why we need the list of blocks to roll forward here.

Reply:
It seems we do, but maybe we could rename `Block` to something else that does  or remove the type synnonim.

```haskell
    | Switch Int (Pushes ds k v)
    -- ^ @Switch k ps@ switche to a different diff sequence by rolling back @k@ elements, 
    -- and then pushing the differences specified by `Pushes`.
```